source/nodejs/adaptivecards-designer/src/card-designer.ts

```diff
@@ -1398,6 +1424,26 @@ export class CardDesigner extends Designer.DesignContext {
         }
     }
 
+    get deviceEmulation(): DeviceEmulation {
+        return this._deviceEmulation;
+    }
+
+    set deviceEmulation(value: DeviceEmulation) {
+        if (this._deviceEmulation !== value) {
```

Comment:
We should only show the picker when it's in preview mode.
but I didn't see you have code to check this condition

Reply:
@licanhua I don't agree with that. The picker is useful as the user designs the card, so they can make quick adjustments after switching from one form factor to another. That is a desirable capability for Teams developers who this feature is designed for.

@thure I assume the selected device setting also applies to the preview when in preview mode?

Reply:
I might be missing some context, is the preview part of the designer ever _not_ in preview mode?

Certainly the idea is to offer this control exactly as often as it is relevant, and based on what I know about the experience here, that is all the time.

Reply:
Do you know how often the user will use templating? If templating is used, this will provide false boundary conditions which doesn't mean too much for developers.

Reply:
> If templating is used, this will provide false boundary conditions

Can you explain this point a bit more? What boundary conditions are you referring to, and how would this provide false ones?

Reply:
@thure you toggle between design and preview modes by clicking the "Preview" button.

Reply:
@licanhua it makes little sense to:
- Let a developer author a card using one specific width
- Then switch to preview mode
- Then switch to a different device emulation just to see that something's off
- Exit preview just to be returned to the previous width they were using to design the card
  - The developer needs to be able to tweak the card while seeing it using the emulation in which it looks off

Besides, Preview mode is basically only needed IF templating is used. If templating is not used and the device emulation selector is only available in preview mode, then the developer may never even see it.

The device emulation selector really has to be available in design mode.

Reply:
For example, if weather card is selected, the "$" part is not the actual rendering until preview mode is clicked

![image](https://user-images.githubusercontent.com/6290692/156048257-d1702209-30e1-45e7-802f-13d5bc42a021.png)

Reply:
As I say above, the selector being available in design mode is crucial when not using templating.

Reply:
I leave this to @kat-y and @JeanRoca

Reply:
Since you are deferring this, I will re-request your review. Please feel free to update your review as needed.

Reply:
From code aspect, it looks good. You can merge it if Kat or J.P. think templating is good

Reply:
Understood, thanks! ✨

Reply:
> Do you know how often the user will use templating? If templating is used, this will provide false boundary conditions which doesn't mean too much for developers.

Hey @licanhua could you explain what you mean by false boundary conditions? I am not sure I understand how the views would impact a developer using templating. 

@thure is this feature available for all host apps or only bot framework / teams?

Reply:
@JeanRoca I've just adjusted the implementation to only apply to hosts that don't specify a card host width in their CSS.